Source/WebCore/dom/Document.cpp

```diff
@@ -3542,7 +3542,9 @@ void Document::setURL(const URL& url)
 
     if (SecurityOrigin::shouldIgnoreHost(newURL))
         newURL.setHostAndPort({ });
-    m_url = WTFMove(newURL);
+    // SecurityContext::securityOrigin may not be initialized at this time if setURL() is called in the constructor, therefore calling topOrigin() is not always safe.
+    auto topOrigin = isTopDocument() && !SecurityContext::securityOrigin() ? SecurityOriginData::fromURL(url) : this->topOrigin().data();
```

Comment:
What if the URL is a blob URL? (which is the base we care about I believe).
I see 2 likely issues:
1. SecurityOriginData::fromURL(url) likely won't be able to extra the origin from a blob URL
2. Even if we addressed the first issue, how would it enforce partitioning?

Imagine this case:
- Frame from origin B (under top level origin A) creates a Blob and gets a URL for it. Then it opens a new window and uses this blob URL as URL for the popup. What's supposed to happen? I imagine the blob should have the "A / B" partitioned origin. Seems like we shouldn't be able to load this Blob in a top level popup? If so, I worry that your code here is not going to achieve this (even if we fix issue 1). Am I missing something?

Reply:
Right, I originally used SecurityOrigin::create(url) for this but I change it in this patch, and that is wrong - we do need SecurityOrigin's logic for handling blob URLs. But, as for (2), there are two options:
1. the popup has "A / B" blob partitioning (therefore has access to the blob data)
2. the popup has "B" blob partitioning (therefore doesn't have access to the blob data)

I was following (2) because creating a top-level navigation with subframe partitioning was conceptually confusing and required more invasive code changes. Following (1) will provide better web compat, but I don't know how common is that use case. If you think we should use (1), then this patch will need changes.

Reply:
I think 2 makes sense. The popup would have "B" origin and thus shouldn't be able to load the blob.

Reply:
Okay, thanks. I also just remembered and realized you are missing some other context. In https://github.com/w3c/FileAPI/issues/153 there is discussion about how to partition blobs, and the general agreement is that cross-origin top-level navigations to a Blob URL should enforce no-opener.